File: patterns/switch-dimmer/rules-js/switch-dimmer.js
const timers = new Map();
const timestamps = new Map();

function switchPressed(id) {
  timestamps.set(id, time.ZonedDateTime.now());
  items.getItem(id).sendCommand("ON");

  const t = actions.ScriptExecution.createTimer(time.ZonedDateTime.now().plusMillis(500), () => {
    items.getItem(id + "LongPress").sendCommand("ON");
    items.getItem(id + "LongPress").sendCommand("OFF");
    const pressedAt = timestamps.get(id);
    if (pressedAt && time.ZonedDateTime.now().isBefore(pressedAt.plusSeconds(5))) {
      timers.get(id).reschedule(time.ZonedDateTime.now().plusMillis(500));
    }
  });
  timers.set(id, t);
}

function switchReleased(id) {
  if (timers.has(id)) {
    timers.get(id).cancel();
    timers.delete(id);
  }
  const pressed = timestamps.get(id);
  timestamps.delete(id);

  if (pressed && time.ZonedDateTime.now().isBefore(pressed.plusMillis(500))) {
    items.getItem(id + "ShortPress").sendCommand("ON");
    items.getItem(id + "ShortPress").sendCommand("OFF");
  }

  items.getItem(id).sendCommand("OFF");
}

rules.JSRule({
  name: "Switch1Events",
  triggers: [triggers.ChannelEventTrigger("enocean:rockerSwitch:xxxxxxxx:xxxxxxxx:rockerswitchA")],
  execute: (event) => {
    switch (event.receivedEvent) {
      case "DIR1_PRESSED":  switchPressed("MySwitch1Up"); break;
      case "DIR1_RELEASED": switchReleased("MySwitch1Up"); break;
      case "DIR2_PRESSED":  switchPressed("MySwitch1Down"); break;
      case "DIR2_RELEASED": switchReleased("MySwitch1Down"); break;
    }
  }
});

File: patterns/switch-dimmer/rules-python/switch_dimmer.py
import threading
from datetime import datetime, timedelta

from openhab import rule, Registry
from openhab.triggers import ChannelEventTrigger

timers = {}
timestamps = {}


def switch_pressed(id_):
    timestamps[id_] = datetime.now().astimezone()
    Registry.getItem(id_).sendCommand("ON")

    def long_press_check():
        Registry.getItem(id_ + "LongPress").sendCommand("ON")
        Registry.getItem(id_ + "LongPress").sendCommand("OFF")
        pressed_at = timestamps.get(id_)
        if pressed_at and datetime.now().astimezone() < pressed_at + timedelta(seconds=5):
            t = threading.Timer(0.5, long_press_check)
            t.start()
            timers[id_] = t

    t = threading.Timer(0.5, long_press_check)
    t.start()
    timers[id_] = t


def switch_released(id_):
    if id_ in timers:
        timers[id_].cancel()
        del timers[id_]
    pressed = timestamps.pop(id_, None)

    if pressed and datetime.now().astimezone() < pressed + timedelta(milliseconds=500):
        Registry.getItem(id_ + "ShortPress").sendCommand("ON")
        Registry.getItem(id_ + "ShortPress").sendCommand("OFF")

    Registry.getItem(id_).sendCommand("OFF")


@rule(triggers=[ChannelEventTrigger("enocean:rockerSwitch:xxxxxxxx:xxxxxxxx:rockerswitchA")])
class Switch1Events:
    def execute(self, module, input):
        event_name = input.get("event")
        if event_name == "DIR1_PRESSED":
            switch_pressed("MySwitch1Up")
        elif event_name == "DIR1_RELEASED":
            switch_released("MySwitch1Up")
        elif event_name == "DIR2_PRESSED":
            switch_pressed("MySwitch1Down")
        elif event_name == "DIR2_RELEASED":
            switch_released("MySwitch1Down")
